bundle/edu.gemini.ags/src/main/scala/edu/gemini/ags/gems/GemsVoTableCatalog.scala

```diff
@@ -0,0 +1,164 @@
+import edu.gemini.catalog.votable.{CatalogException, VoTableClient}
+import edu.gemini.spModel.core.Target.SiderealTarget
+import edu.gemini.spModel.core.{Magnitude, MagnitudeBand, Coordinates}
+import edu.gemini.spModel.gemini.gems.GemsInstrument
+import edu.gemini.spModel.obs.context.ObsContext
+
+import scala.concurrent.{Await, Future}
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.collection.JavaConverters._
+
+import scalaz._
+import Scalaz._
+
+import jsky.util.gui.StatusLogger
+
+/**
+ * Implements GeMS guide star search.
+ * The catalog search will provide the inputs to the analysis phase, which actually assigns guide stars to guiders.
+ * See OT-26
+ */
+object GemsVoTableCatalog {
+  private val DefaultSaturationMagnitude = 0.0
+
+  /**
+   * Searches for the given base position according to the given options.
+   * Multiple queries are performed in parallel in background threads.
+   * This method is synchronous and can be used form the Java side of the OT
+   *
+   * @param obsContext   the context of the observation (needed to adjust for selected conditions)
+   * @param basePosition the base position to search for
+   * @param options      the search options
+   * @param nirBand      optional NIR magnitude band (default is H)
+   * @param timeout      Timeout in seconds
+   * @return list of search results
+   */
+  def search4Java(obsContext: ObsContext, basePosition: Coordinates, options: GemsGuideStarSearchOptions, nirBand: Option[MagnitudeBand], statusLogger: StatusLogger, timeout: Int = 10): java.util.List[GemsCatalogSearchResults] =
+    Await.result(search(obsContext, basePosition, options, nirBand, statusLogger), timeout.seconds).asJava
+
+  /**
+   * Searches for the given base position according to the given options.
+   * Multiple queries are performed in parallel in background threads.
+   *
+   * @param obsContext   the context of the observation (needed to adjust for selected conditions)
+   * @param basePosition the base position to search for
+   * @param options      the search options
+   * @param nirBand      optional NIR magnitude band (default is H)
+   * @return  Future with a list of search results
+   */
+  def search(obsContext: ObsContext, basePosition: Coordinates, options: GemsGuideStarSearchOptions, nirBand: Option[MagnitudeBand], statusLogger: StatusLogger): Future[List[GemsCatalogSearchResults]] = {
+    val criterions = options.searchCriteria(obsContext, nirBand).asScala.toList
+    val inst = options.getInstrument
+
+    val resultSequence = inst match {
+      case GemsInstrument.flamingos2 => searchCatalog(basePosition, criterions, statusLogger)
+      case i                         => searchOptimized(basePosition, criterions, i, statusLogger)
+    }
+
+    // sort on criteria order
+    resultSequence.map(_.toList.sortWith({
+      case (x, y) =>
+        criterions.indexOf(x.criterion) < criterions.indexOf(y.criterion)
```

Comment:
Is it just me, or is doing a call to `indexOf` for each comparison really inefficient? I'm not sure that it matters based on the sizes of `criterions`, but perhaps these indices should be precomputed as this looks like it'll make the sort `2 x^2 log(x)` complexity as opposed to `x log(x)` if they are precomputed.